### PR TITLE
Jetpack connect: Rip out queryObject state

### PR DESCRIPTION
**Pull request by sirreal**
This PR separates the no direct access component from the authorization components.

It removes the page middleware responsible inserting `queryObject` in state.

It validates queryData against a schema and transforms it before passing the required props down to components as required.

It avoids a redirection which effectively strips the query from the URL.

It passes the `redirectAfterAuth` to the plans page via the query rather than state lookup.

Note that some analytics from the authorize components will not be fired when accessing the page directly. This is a slight change from the current behavior. I wouldn't expect that to be a problem cc/ @johnHackworth 

## Testing
1. Try accessing directly https://calypso.live/jetpack/connect/authorize?branch=update/jetpack-connect/auth-form-middleware
1. Ensure you see the no direct access warning 
1. Try a connection flow (logged-in/out) and ensure that it continues to work normally.

Flows:
- [x] Connect starting logged in
- [x] Connect starting logged out
- [x] WooCommerce Services
  * Edit Jetpack plugin to point to `calypso.localhost:3000` (P7rd6c-Me-p2)
  * Install WooCommerce Services plugin and start via the banner at WP Admin
  * Should auto-auth and skip plans
- [x] WooCommerce Wizard
  * Edit Jetpack plugin to point to `calypso.localhost:3000` (P7rd6c-Me-p2)
  * Install WooCommerce plugin, select an option using WooCommerce Services, and connect via banner
  * Should auto-auth and skip plans
- [x] Jetpack Partner NUX flow
  * Instructions at #20025 
- [x] JPO
  * Modify the "from" query param to a string starting with `jpo`
  * Should skip no plans
- [x] SSO (see https://github.com/Automattic/wp-calypso/pull/19956#issuecomment-347258158)
  1. Connect a site
  1. Enable SSO (activate default modules works)
  1. Add a new user with an email associated with an existing WordPress.com account
  1. While connected to WordPress.com the account from the previous step, visit the site `/wp-admin`
  1. Click the SSO button
  1. Make sure you hit the dev site
  1. Verify you're connected.

### CODE REVIEW

**Comment by sirreal** [comment]
Need to work out some kinks with the valid auth form.

**Comment by sirreal** [comment]
> Need to work out some kinks with the valid auth form.

✅  bcd491952506e62c277df49061e6d0c30f7e2fce

**Comment by sirreal** [comment]
After some private conversation with @seear, we've decided to take a slightly different approach here.

* Don't rely on previous middleware to set state
* These are the same concern
* Change this approach to `parseQueryOrFail` 
* schema for queryArgs
* if queryArgs are not ok, we show the no direct access message

**Comment by sirreal** [comment]
redirect without locale if logged in is broken ATM

**Comment by sirreal** [comment]
> redirect without locale if logged in is broken ATM

636e995 fixes this for the fail case (no query params)

For any case including query params, the `getLocaleFromPath` implementation is buggy… off to fix that.

**Comment by sirreal** [comment]
Rebased

**Comment by sirreal** [comment]
Tough rebase 😓

**Comment by sirreal** [comment]
Rebased and updated. Finished removing most of `queryObject` from state, with the exception of client_id.

File: client/jetpack-connect/auth-form-header.jsx [hunk 1]
```diff
@@ -120,15 +118,17 @@ class AuthFormHeader extends Component {
...
 	getSiteCard() {
-		const { jetpackVersion } = this.props;
-		if ( ! versionCompare( jetpackVersion, '4.0.3', '>' ) ) {
+		const { authJpVersion } = this.props;
+		if ( ! versionCompare( authJpVersion, '4.0.3', '>' ) ) {
 			return null;
 		}
 
 		return (
 			<SiteCard
```
Comment (seear) [inline client/jetpack-connect/auth-form-header.jsx]: Might be work quickly checking whether `site-card` needs to exist as a component once we split all the props out of the `queryObject`.

**Comment by sirreal** [comment]
Rebased! Let's test!

**Comment by sirreal** [comment]
I've been through the testing flows in the description myself and haven't encountered any issues.

File: client/jetpack-connect/auth-logged-in-form.jsx [hunk 2]
```diff
@@ -230,50 +232,51 @@ export class LoggedInForm extends Component {
 
 	handleSignOut = () => {
 		const { recordTracksEvent } = this.props;
-		const { queryObject } = this.props.authorizationData;
-		const redirect = addQueryArgs( queryObject, window.location.href );
 		recordTracksEvent( 'calypso_jpc_signout_click' );
-		userUtilities.logout( redirect );
+		userUtilities.logout( window.location.href );
```
Comment (seear) [inline client/jetpack-connect/auth-logged-in-form.jsx]: hehe nice.

Comment (sirreal) [reply client/jetpack-connect/auth-logged-in-form.jsx]: Magical, right? 😀

File: client/jetpack-connect/authorize-form.jsx [hunk 3]
```diff
@@ -10,40 +10,28 @@ import { localize } from 'i18n-calypso';
...
 		// set anonymous ID for cross-system analytics
-		const { authorizationRemoteQueryData } = this.props;
-		if (
-			authorizationRemoteQueryData &&
-			authorizationRemoteQueryData._ui &&
-			'anon' === authorizationRemoteQueryData._ut
-		) {
-			this.props.setTracksAnonymousUserId( authorizationRemoteQueryData._ui );
+		const { authTracksUi, authTracksUt } = this.props;
+		if ( 'anon' === authTracksUt && authTracksUi ) {
```
Comment (seear) [inline client/jetpack-connect/authorize-form.jsx]: I appreciate the better naming here 👍

File: client/jetpack-connect/controller.js [hunk 4]
```diff
@@ -141,39 +132,65 @@ export function connect( context, next ) {
...
+		context.primary = (
+			<JetpackConnectAuthorizeForm
+				path={ context.path }
+				interval={ interval }
+				locale={ locale }
+				authQuery={ transformedQuery }
+			/>
+		);
+	} else {
+		context.primary = <NoDirectAccessError />;
```
Comment (seear) [inline client/jetpack-connect/controller.js]: Nice.

File: client/jetpack-connect/utils.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+		redirectAfterAuth: queryObject.redirect_after_auth || null,
+		siteIcon: queryObject.site_icon || null,
+		siteUrl: queryObject.site_url || null,
+		tracksUi: queryObject._ui || null,
+		tracksUt: queryObject._ut || null,
+		userEmail: queryObject.user_email || null,
+	};
+}
+
+export const authQueryPropTypes = PropTypes.shape( {
```
Comment (seear) [inline client/jetpack-connect/utils.js]: Love the way this keeps the components tidy.

**Comment by seear** [comment]
Code looks fine. Tough to review, but a lot of it is prop changes. I pretty much skipped over the tests.

Going to run through some tests now.

**Comment by seear** [comment]
The testing instructions in this PR are really useful, we might be able to refresh the fieldguide a bit based on this list.

File: client/jetpack-connect/authorize-form.jsx (hunk 3 above)
Comment (sirreal) [reply client/jetpack-connect/authorize-form.jsx]: 😬 I actually missed this in fa2b3e74f9ff4c3a4c95141d8c9254b8fc5a3273, need to fix it up to pull from `authQuery`

Comment (sirreal) [reply client/jetpack-connect/authorize-form.jsx]: ✅ 8271211a8a380bd874684ff85c5c4b344f5f2d75

**Comment by tyxla** [comment]
Thank you for this, amazing work 👍 💯

**Comment by sirreal** [comment]
> The testing instructions in this PR are really useful, we might be able to refresh the fieldguide a bit based on this list.

Updated: PCYsg-dek-p2

**Pull request merged**